Refactoring and

bilabbasi opened the pull request:
The main features added from this PR are as follows:

- a separate Metric class to simplify the FaceMetric, TangentMetric, etc.
- a pooling operation for meshes
- adds scripts to easily run experiments
- adds scripts for visualization of meshes
- the train scripts have been significantly refactors for ease of reuse through out the repo (and not to be used as just a stand alone script)

trachelr commented:
Hi Bilal, thanks a lot for your PR.
Let me test it and do a quick review.

trachelr commented on README.md:
````diff
@@ -67,26 +70,45 @@ out = net(features,mesh.pos,mesh.edge_index,mesh.face.t())
+    ```
+    python preprocess/install_coseg.py --data-dir data/COSEG --ratio 0.85
+    ```
+To use the exact train/test split used in our experiments, add the flag `--splits preprocess/coseg/splits.yaml` to the above command.
````
could you specify that's the default value in the splits argument please?

trachelr commented on environment.yml:
```diff
@@ -7,11 +7,17 @@ dependencies:
     - numpy==1.19.1
     - matplotlib==3.3.0
-    - pytorch==1.5.0
+    - pytorch
```
I think pytorch_geometric is missing, please see pyg package here
https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html

bilabbasi replied:
The pytorch_geometric installation has separate instructions in the README, since it requires some additional stuff. There's a specific note in the README indicating that the environment.yaml doesn't contain installation for pytorch_geometric.

trachelr replied:
my bad, I didn't read that note. I installed pyg version 2.0.4 and py38_torch_1.11.0_cu115
Let it like that if you'd like to keep the same installation version.

trachelr commented on models/mesh_pool.py:
```diff
@@ -0,0 +1,225 @@
+        if in_channels == 0:
+            return None
+        else:
+            return nn.Sequential(Linear(2, 1), nn.LeakyReLU())
```
`nn` is not defined, please use `torch.nn` here

trachelr commented on scripts/visualize_metric.py:
```diff
@@ -1,60 +1,121 @@
+            eigs = metric.eig()[0][:, 0]
+            lambda_max[i, j] = eigs.max()
+            lambda_min[i, j] = eigs.min()
+            # weights[j] = (lambda_max*lambda_min)
```
please remove unnecessary comment

trachelr commented on scripts/visualize_pools.py:
```diff
@@ -0,0 +1,61 @@
+
+        j = 0
+        for (x, v, e, f, cluster) in model.mesh_per_layer:
+            # _,v,e,f,__ = pool(torch.rand(len(v)),v,e,f.type(torch.long).t())
```
same here, please remove these comments if it's not needed anymore...

trachelr commented on preprocess/coseg/convert_labels.py:
```diff
@@ -1,45 +1,55 @@
 import trimesh
+import yaml
+
+config = yaml.safe_load(open("config.yml"))
```
please provide full path to `config.yml` or remove this line

trachelr replied:
@bilabbasi did you saw this comment?

trachelr commented on preprocess/install_human_seg.py:
```diff
@@ -1,122 +1,33 @@
 
-from psbody.mesh import Mesh
-import tmp
+# config = yaml.safe_load(open("config.yml"))
```
please remove this line

trachelr commented on tests/test_compute_face_area_and_angle.py:
```diff
@@ -79,6 +95,9 @@ def test_compute_face_area_and_angle():
+    assert (torch.abs(answer_area - area) < 1e-6).sum() == 4
+    assert (torch.abs(answer_std - std) < 1e-4).sum() == 4
+    assert (torch.abs(answer_angle - angle) < 1e-6).sum() == 4
+    # assert (torch.abs(answer_area - area) < 1e-6).sum() == 4 and (torch.abs(answer_std-std)<1e-6).sum() == 4  and (torch.abs(answer_angle - angle) < 1e-6).sum() == 4
```
same here please

trachelr commented on train/train_args.py:
```diff
@@ -1,40 +1,118 @@
+        action="store_true",
+        help="Indicate whether task is a classification task or not.",
+    )
+    parser.add_argument("layers", nargs="*", default=[16, 32, 64, 128], help="Architecture for MetricConvNet")
```
missing `--` prefix in "layers" here

trachelr commented on scripts/visualize_metric.py:
```diff
@@ -1,60 +1,121 @@
+        pred,
+        fname=join(target_dir, f"{args['dataset']}_{data_idx}_pred.ply"),
+    )
+    pdb.set_trace()
```
Please remove this line as well as the pdb import

trachelr commented on scripts/visualize_metric.py:
```diff
@@ -1,60 +1,121 @@
+        weights = torch.ones(len(v))
+        for j in range(len(v)):
+            metric = m[j].t().mm(m[j])
+            eigs = metric.eig()[0][:, 0]
```
There's a warning here, please use `torch.linalg.eig`

bilabbasi replied:
That's strange; I'm not getting an error. Just to make sure, is your PyTorch up-to-date? I just ran it on my side, and nothing.

trachelr replied:
I have version 1.11

trachelr commented on scripts/visualize_pools.py:
```diff
@@ -0,0 +1,61 @@
+        target_dir = join(log_dir, "samples")
+    os.makedirs(target_dir, exist_ok=True)
+
+    model, args = utils.get_model(log_dir, load_params=True)
```
could you check if the model is an `AutoEncoder` here?
I've got errors with other architectures because it doesnt have the `mesh_per_layer` attribute used below (L40)
Otherwise could you make this script compatible with other kind of models?

trachelr reviewed (commented):
Very nice improvement of the code!! I was able to include pytorch_geometric in the `environment.yaml` file to install it with conda. I installed the CPU version in Ubuntu18.04 (through WSL) but I wasnt able to try GPU support. I've train a model for few epochs in the segmentation task with COSEG and HumanSeg datasets but I wasn't able to get the FAUST dataset for correspondance tasks. The scripts for metric visualisation is working on COSEG but I wasnt able to run the script for pools visualisation. Is it supposed to work with autoencoders only? There's a few pytests errors as well, please see logs in the attached file.
Thanks again for your contribution! 
[test.txt](https://github.com/eidosmontreal/shape-analysis/files/8642015/test.txt)

bilabbasi commented:
Hey Romain, thanks for the feedback. I've addressed a lot of the small things you've mentioned and submitted the changes. For the FAUST installation, can you tell me what your error is? If it persists, then I can just add details on how to download/set up the FAUST dataset.

I've removed the mesh pooling for now, since it's clear it requires a bit more testing before being implemented. 


Thanks for your review so far :)

trachelr commented:
Hello Bilal :)
Thanks for your changes and sorry for the mess-up with the installation... actually I'm running the code in a windows environment. For the FAUST dataset, it's a license issue, you know I'm not allowed to download the dataset... let's skip this part. Did you run the tests? I saw a few errors, don't know if your changes fixed it...
If you're confident with your changes, we can merge.
